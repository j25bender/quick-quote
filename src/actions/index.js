export const addRandomQuote = (randomQuote) => ({
  type: 'ADD_RANDOM_QUOTE',
  randomQuote
});

export const addHomeQuote = (homeQuote) => ({
  type: 'ADD_HOME_QUOTE',
  homeQuote
});

export const addCategoryQuote = (categoryQuote) => ({
  type: 'ADD_CATEGORY_QUOTE',
  categoryQuote
});

export const toggleFavorite = (favoriteQuote) => ({
  type: 'TOGGLE_FAVORITE',
  favoriteQuote
});

export const renderErrorMessage = (error) => ({
  type: 'RENDER_ERROR_MESSAGE',
  error
});
