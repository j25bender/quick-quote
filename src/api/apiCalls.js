import { apiKey } from './apiKey';

export const fetchRandomQuote = async () => {
  try {
    const initialRandomFetch = await fetch('http://quotes.rest/quote/random', {
      method: 'GET',
      headers: {
        'X-TheySaidSo-Api-Secret': apiKey,
        'Accept': 'application/json'
      }
    })
    const randomQuoteJson = await initialRandomFetch.json();
    const quote = randomQuoteJson.contents.quote;
    const author = randomQuoteJson.contents.author;
    const id = randomQuoteJson.contents.id;
    const categories = randomQuoteJson.contents.categories;

    return { quote, author, id, categories }

  } catch(error) {
    throw new Error(`fetchRandomQuote failed to fetch due to: ${error}`)
  }
} 

export const fetchHomeQuote = async () => {
  try {
    const initialMax100Fetch = await fetch('http://quotes.rest/quote/search.json?maxlength=100', {
      method: 'GET',
      headers: {
        'X-TheySaidSo-Api-Secret': apiKey,
        'Accept': 'application/json'
      }
    })
    const max100Json = await initialMax100Fetch.json();
    const quote = max100Json.contents.quote;
    const author = max100Json.contents.author;
    const id = max100Json.contents.id;
    const categories = max100Json.contents.categories;

    return { quote, author, id, categories }

  } catch(error) {
    throw new Error(`fetchHomeQuote failed to fetch due to: ${error}`)
  }
}

export const fetchQuoteCategories = async (category) => {
  try {
    const initialCategoriesFetch = await fetch(`http://quotes.rest/quote/search?category=${category}&minlength=100&maxlength=300&private=false`, {
      method: 'GET',
      headers: {
        'X-TheySaidSo-Api-Secret': apiKey,
        'Accept': 'application/json'
      }
    })
    const categoryQuoteJson = await initialCategoriesFetch.json();
    const quote = categoryQuoteJson.contents.quote;
    const author = categoryQuoteJson.contents.author;
    const id = categoryQuoteJson.contents.id;
    const categories = categoryQuoteJson.contents.categories;

    return { quote, author, id, categories }

  } catch(error) {
    throw new Error(`fetchQuoteCategories failed to fetch due to: ${error}`)
  }
}

export const fetchQuoteOfTheDay = async () => {
  try {
    const initialQuoteOfDayFetch = await fetch('http://quotes.rest/qod.json', {
      method: 'GET',
      headers: {
        'X-TheySaidSo-Api-Secret': apiKey,
        'Accept': 'application/json'
      }
    })
    return initialQuoteOfDayFetch.json();

  } catch(error) {
    throw new Error(`fetchQuoteOfTheDay failed to fetch due to: ${error}`)
  }
}
