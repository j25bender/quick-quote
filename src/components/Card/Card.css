.all-cards {
  display: flex;
  flex-direction: row;
  overflow-y: scroll; }

.quote-container {
  display: flex;
  flex-direction: row; }

.next-button {
  position: absolute;
  background-image: url("../../assets/right-arrow3.png");
  background-size: 40px;
  width: 40px;
  height: 40px;
  background-repeat: no-repeat;
  background-color: transparent;
  border: none;
  right: 30px;
  top: 50%;
  animation: slide 0.5s forwards; }

.favorite-button {
  background-image: url("../../assets/unfavorited.png");
  background-size: 40px;
  width: 40px;
  height: 40px;
  background-repeat: no-repeat;
  background-color: transparent;
  border: none; }

.favorite-button:hover {
  background-image: url("../../assets/hover-favorite.png");
  background-size: 40px;
  width: 40px;
  height: 40px;
  background-repeat: no-repeat;
  background-color: transparent;
  border: none; }
