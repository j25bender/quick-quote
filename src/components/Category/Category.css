@import url("https://fonts.googleapis.com/css?family=Courgette|Kaushan+Script|Special+Elite");
.nonFavorite,
.favorite {
  width: 100vw;
  height: 95vh;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-flow: column wrap;
  line-height: 2;
  font-size: 30px;
  text-shadow: 1px 1px ghostwhite;
  font-weight: bold;
  background-image: url("../../assets/landscape6.jpg");
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover; }

.nonFavorite div,
.favorite div {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-flow: column wrap;
  text-align: center; }

.categories,
.quote,
.author {
  max-width: 80%;
  overflow-y: hidden; }

.nonFavorite button,
.favorite button {
  font-size: 18px; }

div.categories {
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around; }

div.categories p {
  font-size: 20px;
  padding: 5px 5px;
  font-size: 15px;
  overflow-y: hidden; }
