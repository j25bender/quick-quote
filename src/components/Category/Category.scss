@import url('https://fonts.googleapis.com/css?family=Courgette|Kaushan+Script|Special+Elite');

.quote-container {
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around;
  text-align: center;
  font-family: 'Special Elite', cursive;
}

.categories {
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around;
}

.categories p {
  padding: 5px 5px;
  font-size: 15px;
}

.nonFavorite {
  border: 1px black solid;
  width: 300px;
  height: 325;
  padding: 25px;
  margin: 20px;
}

.nonFavorite {
  line-height: 2;
  font-size: 18px;
  background-color: ghostwhite;
}

.nonFavorite button {
  font-size: 18px;
}