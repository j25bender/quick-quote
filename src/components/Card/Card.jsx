import React from 'react';
import './Card.css'
import PropTypes from 'prop-types';

const Card = (props) => {
  const { quote, author = "~ Anonymous", categories } = props.data;
  const cardHtml = <div><h4 className="quote">{ quote }</h4><h5 className="author">{ author }</h5></div>
  if(props) {
    const favoriteCheck = props.data.favorite ? "favorite" : "nonFavorite";
    const categoryList = categories.map( (cat, i) => <p key={ i }>{ cat }</p> )
    return (
      <div>
        <div>
          <div className={ favoriteCheck }>{ cardHtml }</div>
          <div className="categories">{ categoryList }</div>
          <button onClick={ () => props.handleClick(categories) }></button>
        </div>
      </div>
    )
  }
}

Card.propTypes = {
  data: PropTypes.shape({
    quote: PropTypes.string.isRequired,
    author: PropTypes.string,
    id: PropTypes.string.isRequired,
    categories: PropTypes.arrayOf(PropTypes.string)
  }).isRequired,

  handleClick: PropTypes.func.isRequired
};

export default Card;