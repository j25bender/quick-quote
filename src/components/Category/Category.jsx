import React, { Component } from 'react';
import './Category.css';
import { addHomeQuote, addRandomQuote, addCategoryQuote } from '../../actions';
import { fetchHomeQuote, fetchRandomQuote, fetchQuote } from '../../api/apiCalls';
import Card from '../Card/Card';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

export class Category extends Component {

  componentDidMount() {
    const pathnameProp = this.props.location.pathname;    
    const cleanCategory = pathnameProp.slice(1);
    this.fetchAndDispatch(cleanCategory);
  }

  componentWillReceiveProps(nextProps) {
    // double check this is working as expected
    // test is props.location nextprops match dont match
    console.log('this.props.location.key', this.props, Date.now())
    console.log('nextProps.location.key', nextProps, Date.now())
    console.log('Do this,props.location.key !== nextProps.location.key?', this.props.location.key !== nextProps.location.key)
    if(this.props.location.key !== nextProps.location.key) {
      const pathnameProp = nextProps.location.pathname;
      const cleanCategory = pathnameProp.slice(1);
      this.fetchAndDispatch(cleanCategory);
    }
  }

  handleClick = async (categories) => {
    // test it calls fetch quote cat and add category quote
    const randomCategory = categories.length !== 1 ? categories[ Math.floor( Math.random() * categories.length ) ] : 'funny';
    try {
      const categoryQuoteToDispatch = await fetchQuote(randomCategory);
      this.props.addCategoryQuote(categoryQuoteToDispatch);      
    } catch(error) {
      throw new Error(`handleClick failed to fetch: ${error}`)
    }
  }

  async fetchAndDispatch(category) {
    // test given a category the correct fetch and add action gets called and errors
    const { homeQuote, randomQuote, categoryQuote } = this.props;

    if(category === 'home') {
      try {
        const homeQuoteToDispatch = await fetchHomeQuote();
        this.props.addHomeQuote(homeQuoteToDispatch);
        return [ ...homeQuote, homeQuoteToDispatch ];
      } catch(error) {
        throw new Error(`fetchHomeQuote in fetchAndDispatch failed: ${error}`);
      }

    } else if (category === 'random') {
      try {
        const randomQuoteToDispatch = await fetchRandomQuote();
        this.props.addRandomQuote(randomQuoteToDispatch);
        return [ ...randomQuote, randomQuoteToDispatch ];
      } catch(error) {
        throw new Error(`fetchRandomQuote in fetchAndDispatch failed: ${error}`)
      }
      
    } else {
      try {
        const categoryQuoteToDispatch = await fetchQuote(category);
        this.props.addCategoryQuote(categoryQuoteToDispatch);
        return [ ...categoryQuote, categoryQuoteToDispatch ];
      } catch(error) {
        throw new Error(`fetchQuote in fetchAndDispatch failed: ${error}`)
      }
    }
  }

  renderCards = () => {
    // test w snapshot passing in diff quotes
    const { homeQuote, randomQuote, categoryQuote } = this.props;
    const pathnameProp = this.props.location.pathname;
    let quoteToUse
    if(pathnameProp.includes('random')) {
      quoteToUse = randomQuote
    } else if(pathnameProp.includes('home')) {
      quoteToUse = homeQuote
    } else {
      quoteToUse = categoryQuote
    }
    const cardsToRender = quoteToUse.map( quote => <Card key={ quote.id } data={ quote } handleClick={ this.handleClick } /> )
    return cardsToRender
  }
  
  render() {
    return (
      <div>
        { this.renderCards() }
      </div>
    )
  }
}

Category.propTypes = {
  addHomeQuote: PropTypes.func,
  addRandomQuote: PropTypes.func,
  addCategoryQuote: PropTypes.func,

  homeQuote: PropTypes.arrayOf(PropTypes.shape({
    quote: PropTypes.string.isRequired,
    author: PropTypes.string,
    id: PropTypes.string.isRequired,
    categories: PropTypes.arrayOf(PropTypes.string)
  })),

  randomQuote: PropTypes.arrayOf(PropTypes.shape({
    quote: PropTypes.string.isRequired,
    author: PropTypes.string,
    id: PropTypes.string.isRequired,
    categories: PropTypes.arrayOf(PropTypes.string)
  })),

  categoryQuote: PropTypes.arrayOf(PropTypes.shape({
    quote: PropTypes.string.isRequired,
    author: PropTypes.string,
    id: PropTypes.string.isRequired,
    categories: PropTypes.arrayOf(PropTypes.string)
  })),
};

export const mapStateToProps = (state) => ({
  homeQuote: state.homeQuote,
  randomQuote: state.randomQuote,
  categoryQuote: state.categoryQuote
});

export const mapDispatchToProps = (dispatch) => ({
  addHomeQuote: (homeQuoteToDispatch) => dispatch(addHomeQuote(homeQuoteToDispatch)),
  addRandomQuote: (randomQuoteToDispatch) => dispatch(addRandomQuote(randomQuoteToDispatch)),
  addCategoryQuote: (categoryQuoteToDispatch) => dispatch(addCategoryQuote(categoryQuoteToDispatch))
});

export default connect(mapStateToProps, mapDispatchToProps)(Category);