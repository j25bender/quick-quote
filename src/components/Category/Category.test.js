import React, { Component } from 'react';
import { addHomeQuote, addRandomQuote, addCategoryQuote } from '../../actions';
import { Category, mapStateToProps, mapDispatchToProps } from './Category';
import Card from '../Card/Card';
import { fetchRandomQuote, fetchHomeQuote, fetchQuote } from '../../api/apiCalls';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const fakeStore = (state) => {
  return {
    dispatch: () => {},
    getState: () => {
      return {...state}
    }
  }
}

let wrapper;
let mockfetchAndDispatch;
let mockAddCategoryQuote;
let mockQuoteData = {quote: "Retirement is the ugliest word in the language.", 
                          author: "Ernest Hemingway", 
                          id: "hqV6ZjgFus0xQ58WPhH5hQeF", 
                          categories: Array(1)
                         };

beforeEach( () => {
  mockfetchAndDispatch = jest.fn();
  mockAddCategoryQuote = jest.fn();
  window.fetch = jest.fn().mockImplementation( (url) => {
    return Promise.resolve({
      status: 200,
      json: () => Promise.resolve({contents: mockQuoteData})
    })
  })
  wrapper = shallow(<Category location={ {pathname: "/life", key: 'twms5s'} }
                              categoryQuote={ [mockQuoteData] }
                              addCategoryQuote={ mockAddCategoryQuote } />)
})

describe('Category Container Tests', () => {
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should call addCategoryQuote when Category container mounts', () => {
    expect(wrapper.instance().props.addCategoryQuote).toHaveBeenCalled();
  })
  
  it('should update with props.location.key do not equal nextProps.location.key', () => {
    wrapper.setProps({ location: {pathname: "/life", key: 'twms5s'}});
    wrapper.setProps({ nextProps: {location: {pathname: "/life", key: 'twms5s'} }});
    expect(wrapper.instance().props.addCategoryQuote).toHaveBeenCalled()
    wrapper.setProps({ location: {pathname: "/life", key: 'twms5s'}});
    wrapper.setProps({ nextProps: {location: {pathname: "/life", key: 'manana'} }});
    expect(wrapper.instance().props.addCategoryQuote).not.toHaveBeenCalled()
  })
})
