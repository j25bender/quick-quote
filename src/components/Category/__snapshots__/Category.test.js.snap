// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category Container Tests should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Category
    addCategoryQuote={[Function]}
    categoryQuote={
        Array [
            Object {
              "author": "Ernest Hemingway",
              "categories": Array [
                undefined,
              ],
              "id": "hqV6ZjgFus0xQ58WPhH5hQeF",
              "quote": "Retirement is the ugliest word in the language.",
            },
          ]
    }
    location={
        Object {
            "key": "twms5s",
            "pathname": "/life",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Card
          data={
                    Object {
                              "author": "Ernest Hemingway",
                              "categories": Array [
                                undefined,
                              ],
                              "id": "hqV6ZjgFus0xQ58WPhH5hQeF",
                              "quote": "Retirement is the ugliest word in the language.",
                            }
          }
          handleClick={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "hqV6ZjgFus0xQ58WPhH5hQeF",
        "nodeType": "function",
        "props": Object {
          "data": Object {
            "author": "Ernest Hemingway",
            "categories": Array [
              undefined,
            ],
            "id": "hqV6ZjgFus0xQ58WPhH5hQeF",
            "quote": "Retirement is the ugliest word in the language.",
          },
          "handleClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Card
            data={
                        Object {
                                    "author": "Ernest Hemingway",
                                    "categories": Array [
                                      undefined,
                                    ],
                                    "id": "hqV6ZjgFus0xQ58WPhH5hQeF",
                                    "quote": "Retirement is the ugliest word in the language.",
                                  }
            }
            handleClick={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "hqV6ZjgFus0xQ58WPhH5hQeF",
          "nodeType": "function",
          "props": Object {
            "data": Object {
              "author": "Ernest Hemingway",
              "categories": Array [
                undefined,
              ],
              "id": "hqV6ZjgFus0xQ58WPhH5hQeF",
              "quote": "Retirement is the ugliest word in the language.",
            },
            "handleClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
